docs/variance-defaulting.md

```diff
@@ -0,0 +1,375 @@
+`T.untyped`? It's because we can't get away with using `upper` nor `lower`.
+
+Consider this setup:
+
+```ruby
+class Parent; end
+class Child < Parent; end
+```
+
+In this setup, `T.class_of(Child) <: T.class_of(Parent)`, which is what users
+expect. Now consider if we made `Parent` and `Child` generic:
+
+```ruby
+class Parent
+  extend T::Generic
+  Elem = type_template { {upper: Numeric} }
+end
+class Child < Parent
+  extend T::Generic
+  Elem = type_template { {upper: Integer} }
+end
+```
+
+It no longer makes sense to ask whether `T.class_of(Child) <:
+T.class_of(Parent)`, because those are bare generics. But both `upper` and
+`lower` are hard to stomach as defaults.
+
+Consider if we pick `upper`. Then we'd have that:
+
+- `Parent` has type `T.class_of(Parent)[Numeric]` and
+- `Child` has type `T.class_of(Child)[Integer]`
+
+`T.class_of(Child)[Integer]` is not a subtype of `T.class_of(Parent)[Numeric]`
+because `Elem` is invariant! This would be the sound thing to do: it's possible
+to draw a contradiction if we assume that `Child` is an instance of
+`T.class_of(Parent)[Numeric]`:
+
+```ruby
+# typed: strong
+class Module; include T::Sig; end
+
+class Parent
+  extend T::Generic
+  Elem = type_template { {upper: Numeric} }
+
+  sig {overridable.params(x: Elem).void}
+  def self.example(x); end
+end
+
+class Child < Parent
+  extend T::Generic
+  Elem = type_template { {upper: Integer} }
+
+  sig {params(x: Integer).void}
+  def self.takes_integer(x); end
+
+  sig {override.params(x: Elem).void}
+  def self.example(x)
+    takes_integer(x)
+    # ^ our program passes a float to takes_integer without using T.unsafe,
+    # and with no errors
+  end
+end
+
+sig {params(parent_class: T.class_of(Parent)).void}
+#           ^ defaults to T.class_of(Parent)[Numeric]
+def main(parent_class)
+  parent_class.example(0.0)
+  #                    ^^^ a valid numeric, but not an Integer!
+end
+
+main(Child)
+```
+
+So maybe `upper` was the wrong thing to pick? But if we pick `lower`, two things
+happen:
+
+- `T.class_of(Parent)` defaults to `T.class_of(Parent)[T.noreturn]`,
+  which means we can't even call `parent_class.example`
+- We could make a similar contradiction by rewriting the example to use `Elem`
+  in `returns` instead of `params`.
+
+So maybe we pick `upper` when defaulting in `params`, and `lower` when
+defaulting in `returns`? Implementation complexity aside (we'd have to thread
+input/output position (polarity) through all calls to `externalType`, which would
+mean we couldn't cache it as easily as now), we'd still have the problem that
+most users expect `Child` to be a valid `T.class_of(Parent)`.
+
+So Sorbet gives up and accepts the unsoundness by simply defaulting to
+`T.untyped`.
+
+## Be super careful with unfixed, invariant type members
+
+The takeaway? Type members in well-typed code should always either:
+
+- be fixed, and accept the limitations that come with that (basically: makes the
+  class final in practice), or
```

Comment:
This "final in practice" aside seems to come out of nowhere.  How does this happen?

Reply:
it's subtle, I added a footnote